mlrun/serving/server.py

```
@@ -605,14 +608,45 @@ async def async_execute_graph(
                 f"(status='{task_state}')"
             )
 
-    server.graph = add_system_steps_to_graph(
-        server.project,
-        copy.deepcopy(server.graph),
-        spec.get("track_models"),
-        context,
-        spec,
-        pause_until_background_task_completion=False,  # we've already awaited it
-    )
+    mm_enabled = True
+    df = data.as_df()
+
+    if df.empty:
+        context.logger.warn("Job terminated due to empty inputs (0 rows)")
+        return []
+
+    if timestamp_column:
+        context.logger.info(f"Sorting dataframe by {timestamp_column}")
+        df[timestamp_column] = pd.to_datetime(  # in case it's a string
+            df[timestamp_column]
+        )
+        df.sort_values(by=timestamp_column, inplace=True)
+        if len(df) >= 2:
+            start_time = df[timestamp_column].iloc[0]
+            end_time = df[timestamp_column].iloc[-1]
+            time_range = end_time - start_time
+            start_time = start_time.isoformat()
+            end_time = end_time.isoformat()
+            # TODO: tie this to the controller's base period
+            if time_range > pd.Timedelta("10000h"):
+                context.logger.warn(
+                    f"Dataframe time range is too long: {time_range}. Job will not be tracked!"
+                )
+                mm_enabled = False
```

I think that if `track_models=True` and the time range is to long we should raise an exception.